






<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="last-modified" content="2024-10-17 02:23:08 -0400">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- meta "search-domain" used for google site search function google_search() -->
    <meta name="search-domain" value="">
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/lesson.css" />
    <link rel="stylesheet" type="text/css" href="../assets/css/syntax.css" />
    
    <link rel="license" href="#license-info" />

    



    <!-- Favicons for everyone -->
    <link rel="apple-touch-icon-precomposed" sizes="57x57" href="../assets/favicons/swc/apple-touch-icon-57x57.png" />
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="../assets/favicons/swc/apple-touch-icon-114x114.png" />
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="../assets/favicons/swc/apple-touch-icon-72x72.png" />
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="../assets/favicons/swc/apple-touch-icon-144x144.png" />
    <link rel="apple-touch-icon-precomposed" sizes="60x60" href="../assets/favicons/swc/apple-touch-icon-60x60.png" />
    <link rel="apple-touch-icon-precomposed" sizes="120x120" href="../assets/favicons/swc/apple-touch-icon-120x120.png" />
    <link rel="apple-touch-icon-precomposed" sizes="76x76" href="../assets/favicons/swc/apple-touch-icon-76x76.png" />
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../assets/favicons/swc/apple-touch-icon-152x152.png" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-196x196.png" sizes="196x196" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-96x96.png" sizes="96x96" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-16x16.png" sizes="16x16" />
    <link rel="icon" type="image/png" href="../assets/favicons/swc/favicon-128.png" sizes="128x128" />
    <meta name="application-name" content="Software Carpentry - Single Cell RNAseq"/>
    <meta name="msapplication-TileColor" content="#FFFFFF" />
    <meta name="msapplication-TileImage" content="../assets/favicons/swc/mstile-144x144.png" />
    <meta name="msapplication-square70x70logo" content="../assets/favicons/swc/mstile-70x70.png" />
    <meta name="msapplication-square150x150logo" content="../assets/favicons/swc/mstile-150x150.png" />
    <meta name="msapplication-wide310x150logo" content="../assets/favicons/swc/mstile-310x150.png" />
    <meta name="msapplication-square310x310logo" content="../assets/favicons/swc/mstile-310x310.png" />


    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
	<script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
	<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
	<![endif]-->

  <title>
  Introduction to Single Cell RNA-seq &ndash; Single Cell RNAseq
  </title>  

  </head>
  <body>

    








    <div class="container">
      
















  
  










<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      
      
      <a href="https://software-carpentry.org" class="pull-left">
        <img class="navbar-logo" src="../assets/img/swc-icon-blue.svg" alt="Software Carpentry logo" />
      </a>
      

      
      <a class="navbar-brand" href="../index.html">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">

	
        <li><a href="../CODE_OF_CONDUCT.html">Code of Conduct</a></li>

        
	
        <li><a href="../setup.html">Setup</a></li>

        
        
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Episodes <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            
            <li><a href="../01-introduction/index.html">Introduction to Single Cell RNA-seq</a></li>
            
            
            <li><a href="../02-Experimental-Considerations/index.html">Experimental Considerations</a></li>
            
            
            <li><a href="../03-Overview-scRNA-seq-Data/index.html">Overview of scRNA-seq Data</a></li>
            
            
            <li><a href="../04-Quality-Control/index.html">Quality Control of scRNA-Seq Data</a></li>
            
            
            <li><a href="../05-Common-Analyses/index.html">Common Analyses</a></li>
            
            
            <li><a href="../06-Biology-Driven-Analyses/index.html">Biology Driven Analyses of scRNA-Seq</a></li>
            
            
            <li><a href="../07-Analyzing-Your-Data/index.html">Analyzing Your Data</a></li>
            
            
            <li><a href="../08-Future-Directions/index.html">Future Directions</a></li>
            
	    <li role="separator" class="divider"></li>
            <li><a href="../aio/index.html">All in one page (Beta)</a></li>
          </ul>
        </li>
        
	

	
	
        <li class="dropdown">
          <a href="../" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
          <ul class="dropdown-menu">
            <li><a href="../reference.html">Reference</a></li>
            
            
            
              <li><a href="../about/index.html">About</a></li>
            
            
            
            
              <li><a href="../discuss/index.html">Discussion</a></li>
            
            
            
            
              <li><a href="../figures/index.html">Figures</a></li>
            
            
            
            
              <li><a href="../guide/index.html">Instructor Notes</a></li>
            
            
          </ul>
        </li>
	

	
        <li><a href="../LICENSE.html">License</a></li>
	
	<li><a href="/edit//_episodes/01-introduction.md" data-checker-ignore>Improve this page <span class="glyphicon glyphicon-pencil" aria-hidden="true"></span></a></li>
	
      </ul>
      <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form>
    </div>
  </div>
</nav>



















  
  











<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../"><span class="glyphicon glyphicon-menu-up" aria-hidden="true"></span><span class="sr-only">lesson home</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
    <h3 class="maintitle"><a href="../">Single Cell RNAseq</a></h3>
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../02-Experimental-Considerations/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>

<article>
<div class="row">
  <div class="col-md-1">
  </div>
  <div class="col-md-10">
    <h1 class="maintitle">Introduction to Single Cell RNA-seq</h1>
  </div>
  <div class="col-md-1">
  </div>
</div>












<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 45 min
      <br/>
      <strong>Exercises:</strong> 10 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>What is single cell RNA-seq?</p>
</li>
	
	<li><p>What is the difference between bulk RNA-seq and single cell RNA-seq?</p>
</li>
	
	<li><p>How do I choose between bulk RNA-seq and single cell RNA-seq</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Describe the overall experimental process of generating single-cell transcriptomes.</p>
</li>
	
	<li><p>Describe instances where bulk vs. single cell RNA-Seq methods are uniquely appropriate.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<h2 id="brief-overview-of-single-cell-transcriptomics-technology">Brief overview of single cell transcriptomics technology</h2>

<p>Single cell RNA-sequencing (scRNA-Seq) is a method of quantifying transcript 
expression levels in individual cells. scRNA-Seq technology can take on many 
different forms and this area of research is rapidly evolving. In 2022, the 
most widely used systems for performing scRNA-Seq involve separating cells 
and introducing them into a microfluidic system which performs the chemistry 
on each cell individually (droplet-based scRNA-Seq).</p>

<p>In this workshop we will primarily focus on the 10X Genomics technology. 
10X Genomics is a market leader in the single cell space and was among the 
first technologies that made it feasible to profile thousands of cells 
simultaneously. Single cell technology is changing rapidly and it is not 
clear whether any other companies will be able to successfully challenge 
10X’s dominance in this space.</p>

<p>The steps in droplet scRNA-Seq are:</p>

<ol>
  <li>Cell isolation:
    <ul>
      <li>If the cells are part of a tissue, the cells are disaggregated using 
collagenase or other reagents. The specifics of this protocol can vary 
greatly due to differences between tissues that are biological in nature. 
If the cells are in culture or suspension, they may be used as-is.</li>
    </ul>
  </li>
  <li>Assess cell viability.
    <ul>
      <li>If scRNA-Seq is being performed on fresh tissue, the cells are usually 
 checked for viability. We want “happy” cells loaded into the machine. 
 We might hope for &gt;90% viable and set a minimum threshold of &gt;70%, 
 although these numbers can vary greatly depending on the experiment.</li>
    </ul>
  </li>
  <li>Droplet formation inside instrument:
    <ul>
      <li>Using a microfluidic system, each cell is suspended in a nanoliter-size 
 droplet along with a barcoded primer bead. The cells are kept separate 
 from each other in an oil/water emulsion.</li>
    </ul>
  </li>
  <li>Cell lysis, generating complementary DNA (cDNA):
    <ul>
      <li>The cells are lysed in each droplet. Each cell was already encapsulated 
 with a barcoded primer bead which has a primer specific to that cell. 
 Often a poly-d(T) primer is used to prime the poly(A) tail of mRNA. 
 Complementary DNA is transcribed and correct primers are added for 
 Illumina sequencing.</li>
    </ul>
  </li>
  <li>Library generation:
    <ul>
      <li>Oil is removed to destroy droplets and homogenize the mixture. cDNA 
 is amplified by PCR. The cDNA is sequenced on any Illumina machine. 
 Sequencing should be paired-end – one read contains cell and molecule 
 barcodes, while the other read contains the target transcript that was 
 captured. We will discuss the sequence files in more detail in a later
 lesson. Any service provider who generates scRNA-Seq data for you 
 should know how to properly set up in Illumina run to get you the 
 data you need.</li>
    </ul>
  </li>
</ol>

<p><img src="../fig/Zhang_etal_2017_NatComm_Fig1_GEMs.jpg" alt="Single cell library prep" /></p>

<h2 id="what-is-scrna-seq-useful-for">What is scRNA-Seq useful for?</h2>

<p>Single cell RNA-Seq is a new technology and its uses are limited only by your 
imagination! A few examples of problems that have been addressed using 
scRNA-Seq include:</p>
<ul>
  <li>developmental studies &amp; studies of cellular trajectories.</li>
  <li>detailed tissue atlases.</li>
  <li>characterization of tumor clonality.</li>
  <li>definition of cell-type specific transcriptional responses 
 (e.g. T-cell response to infection)</li>
  <li>profiling of changes in cell state (i.e. homeostasis vs. response state)</li>
  <li>a variety of different types of CRISPR screens</li>
</ul>

<h2 id="comparing-and-contrasting-scrna-seq-with-bulk-rna-seq">Comparing and contrasting scRNA-Seq with bulk RNA-Seq</h2>

<p>Bulk RNA-Seq and single cell RNA-Seq are related in that they both assess 
transcription levels by sequencing short reads, but these two technologies 
have a variety of differences. Neither technology is always better. The 
approach that one might use should depend upon the information one hopes 
to gather.</p>

<p>Consider the following points when assessing the differences between 
the technologies and choosing which to utilize for your own experiment:</p>
<ul>
  <li>Tissues are heterogeneous mixtures of diverse cell types. Bulk RNA-Seq 
 data consists of average measures of transcripts expressed across many 
 different cell types, while scRNA-Seq data is cell-type resolved.</li>
  <li>Bulk RNA-Seq data may not be able to distinguish between changes in gene 
 expression versus changes in tissue cell composition.</li>
  <li>Bulk RNA-Seq allows for much higher sequencing coverage for each gene and 
 often captures more total genes.</li>
  <li>Bulk RNA-Seq allows for better isoform detection due to the higher 
 sequencing depth and relatively uniform coverage across transcripts 
 (vs. a typical 3’ bias in scRNA-Seq).</li>
  <li>Genes without poly-A tail (e.g. some noncoding RNAs) might not be detected 
 in scRNA-Seq, but can be reliably assessed using bulk RNA-Seq.</li>
</ul>

<p><img src="../fig/singlecell-vs-bulk-image.png" alt="Single Cell vs. Bulk RNA-seq" />
https://www.10xgenomics.com/blog/single-cell-rna-seq-an-introductory-overview-and-tools-for-getting-started</p>

<blockquote class="challenge">
  <h2 id="challenge">Challenge</h2>
  <p>For each of these scenarios, choose between using bulk RNA-Seq and scRNA-Seq 
to address your problem.</p>

  <p>Differentiation of embryonic stem cells to another cell type</p>
  <blockquote class="solution">
    <h2 id="solution">Solution</h2>
    <p>You would likely find single cell RNA-Seq most powerful in this situation 
since the cells are differentiating along a continuous transcriptional gradient.</p>
  </blockquote>

  <p>Studying aging with a specific focus on the senescence-involved (e.g. <a href="https://pubmed.ncbi.nlm.nih.gov/15582292/">ref</a>) gene <em>Cdkn2a</em></p>
  <blockquote class="solution">
    <h2 id="solution-1">Solution</h2>
    <p>Since you are interested in the expression of a single gene (<em>Cdkn2a</em>), 
bulk RNA-seq may be a better choice because of the greater sequencing depth 
and the potenial to identify isoforms.</p>
  </blockquote>

  <p>Studying variation in vaccine response by profiling peripheral blood 
mononuclear cells (PBMCs) –</p>
  <blockquote class="solution">
    <h2 id="solution-2">Solution</h2>
    <p>Since you are likely interested in gene expression within specific cell 
types, single cell RNA-seq may be a better choice because you will be able to 
quantify cell proportions and cell-specific gene expression.</p>
  </blockquote>

  <p>Doing functional genomics in a non-model species</p>
  <blockquote class="solution">
    <h2 id="solution-3">Solution</h2>
    <p>Non-model organsms may not have a well-developed reference genome or 
transcript annotation. Thus, you may need to use tools which perform de-novo 
transcript assembly and then align your reads to that custom transcriptome. 
De-novo transcript assembly requires greater sequencing depth, which 
single-cell RNA-Seq may not provide. Therefore we would recommend using bulk 
RNA-Seq in this situation.</p>
  </blockquote>

  <p>Studying micro RNAs</p>
  <blockquote class="solution">
    <h2 id="solution-4">Solution</h2>
    <p>MicroRNAs are not currently assayed by most scRNA-Seq technologies. 
Thus bulk RNA-Seq, with an enrichment for small RNAs, would be the better 
choice here.</p>
  </blockquote>

  <p>Performing gene expression quantitative trait locus (eQTL) mapping</p>
  <blockquote class="solution">
    <h2 id="solution-5">Solution</h2>
    <p>You may want good estimates of transcript abundance in your tissue of 
interest, so bulk RNA-Seq may be a good choice here. However, you may be able 
to aggregate scRNA-Seq expression for each cell type and perform eQTL mapping. 
Therefore both technologies could be informative!</p>
  </blockquote>
</blockquote>

<h2 id="single-cell-data-modalities">Single cell data modalities</h2>

<p>There are several different modalities by which one can gather data on the 
molecular properties of single cells. 10X Genomics currently offers the 
following reliable assays:</p>

<ul>
  <li>RNA-seq - assess gene expression in single cells. A sub-group of single cell 
 gene expression is single nucleus gene expression, which is often used on 
 frozen samples or for tissues where preparing a single cell suspension is 
 difficult or impossible (e.g. brain).</li>
  <li>ATAC-seq - assess chromatin accessibility in single cells</li>
  <li>“Multiome” - RNA + ATAC in the same single cells</li>
  <li>Immune repertoire profiling - assess clonality and antigen specificity of 
 adaptive immune cells (B and T cells)</li>
  <li>CITE-Seq - assess cell surface protein levels</li>
  <li>Spatial transcriptomics/proteomics - assess gene expression and/or protein 
 abundance with near-single cell resolution</li>
</ul>

<h2 id="focus-of-this-course-10x-genomics-mouse-scrna-seq">Focus of this course: 10X Genomics mouse scRNA-Seq</h2>

<p>While there are several commercially available scRNA-seq technologies, this 
course will focus on data generated by the 10X Genomics platform. This 
technology is widely used at JAX, and it is flexible, reliable, and relatively
cost-efficient. We will focus on a mouse data set, although most of the 
techniques that you will learn in this workshop apply equally well to other 
species. Each data modality has its own strengths and weaknesses; in this 
course we will focus on transcriptomics.</p>

<h2 id="brief-overview-of-instrumentation-and-library-preparation-choices-available-from-10x-genomics">Brief overview of instrumentation and library preparation choices available from 10X Genomics</h2>

<p>10X Genomics offers a variety of options for profiling single cells. Here we 
give a brief overview of some of their offerings to illustrate what is available.</p>

<p>To profile gene expression, users can choose from several options:</p>
<ul>
  <li>3’ gene expression – the “usual” option, amplifies from the 3’ end of 
 transcripts</li>
  <li>5’ gene expression – another option for profiling gene expression that
 captures from the 5’ end of the transcript. See <a href="https://kb.10xgenomics.com/hc/en-us/articles/360000939852-What-is-the-difference-between-Single-Cell-3-and-5-Gene-Expression-libraries-">this link</a>
 for some information on 3’ vs 5’ gene expression</li>
  <li>“targeted” gene expression – focus on a smaller number of genes that are of 
 particular interest</li>
</ul>

<p>One can also choose to profile gene expression on different instruments. The 
“workhorse” <strong>10X Chromium</strong> machine is used for profiling gene expression of 
up to eight samples totaling 25k-100k cells. The newer Chromium X is a higher 
throughput instrument which can profile up to 16 samples and up to one million 
cells. There are also variations in the kits that can be purchased to perform 
library preparation. In general 10X Genomics does a fairly good job of 
continuously improving the chemistry such that the data quality continues to 
improve. More information is available from 
<a href="https://www.10xgenomics.com/products/single-cell-gene-expression">10X Genomics</a>.</p>

<p>10X Genomics also produces a suite of other instrumentation for applications 
such as spatial transcriptomics (Visium) or in situ profiling (Xenium), which 
we will not cover in this course.</p>






<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>Single cell methods excel in defining cellular heterogeneity and profiling cells differentiating along a trajectory.</p>
</li>
    
    <li><p>Single cell RNA-Seq data is sparse and cannot be analyzed using standard bulk RNA-Seq approaches.</p>
</li>
    
    <li><p>Bulk transcriptomics provides a more in-depth portrait of tissue gene expression, but scRNA-Seq allows you to distinguish between changes in cell composition vs gene expression.</p>
</li>
    
    <li><p>Different experimental questions can be investigated using different single-cell sequencing modalities.</p>
</li>
    
  </ul>
</blockquote>

</article>


















  
  











<div class="row">
  <div class="col-xs-1">
    <h3 class="text-left">
      
      <a href="../"><span class="glyphicon glyphicon-menu-up" aria-hidden="true"></span><span class="sr-only">lesson home</span></a>
      
    </h3>
  </div>
  <div class="col-xs-10">
    
  </div>
  <div class="col-xs-1">
    <h3 class="text-right">
      
      <a href="../02-Experimental-Considerations/index.html"><span class="glyphicon glyphicon-menu-right" aria-hidden="true"></span><span class="sr-only">next episode</span></a>
      
    </h3>
  </div>
</div>


      
      






<footer>
  <div class="row">
    <div class="col-md-6 license" id="license-info" align="left">
	
	Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2018–2024
	by <a href="https://carpentries.org/">The Carpentries</a>
        <br>
        Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2016–2018
	by <a href="https://software-carpentry.org">Software Carpentry Foundation</a>
	
    </div>
    <div class="col-md-6 help-links" align="right">
	
	<a href="/edit//_episodes/01-introduction.md" data-checker-ignore>Edit on GitHub</a>
	
	/
	<a href="/blob//CONTRIBUTING.md" data-checker-ignore>Contributing</a>
	/
	<a href="/">Source</a>
	/
	<a href="/blob//CITATION" data-checker-ignore>Cite</a>
	/
	<a href="mailto:team@carpentries.org">Contact</a>
    </div>
  </div>
  <div class="row">
    <div class="col-md-12" align="center">
      Using <a href="https://github.com/carpentries/styles/">The Carpentries style</a>
      version <a href="https://github.com/carpentries/styles/releases/tag/v9.5.3">9.5.3</a>.
    </div>
  </div>
</footer>

      
    </div>
    
<script src="../assets/js/jquery.min.js"></script>
<script src="../assets/js/bootstrap.min.js"></script>
<script src="../assets/js/lesson.js"></script>


<!-- Matomo -->
<script>
  var _paq = window._paq = window._paq || [];
  /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
  _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
  _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io"]]);
  _paq.push(["setDoNotTrack", true]);
  _paq.push(["disableCookies"]);
  _paq.push(['trackPageView']);
  _paq.push(['enableLinkTracking']);
  (function() {
    var u="https://carpentries.matomo.cloud/";
    _paq.push(['setTrackerUrl', u+'matomo.php']);
    _paq.push(['setSiteId', '1']);
    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
    g.async=true; g.src='//cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
  })();
</script>
<!-- End Matomo Code -->


<script src="../assets/js/anchor.min.js"></script>
<script>
    anchors.add();
</script>

  </body>
</html>
